### PR TITLE
WFLY-5179 Some tests are unable to init arquillian protocol with security manager

### PR DESCRIPTION
**Pull request by OndrejKotek**
Fix for https://issues.jboss.org/browse/WFLY-5179

Fixed test cases avoid using ManagementClient when running in non-RunAsClient mode.
EjbLocalRefInjectionTestCase and EnvEntryInjectionTestCase were split -- no permissions needed.
AuthenticationTestCase and XTS TransactionalTestCase have new permissions and comments.

### CODE REVIEW

**Comment by bstansberry** [comment]
@tomjenkinson Tom, can someone more familiar with xts look at the permissions being assigned to the deployments here to confirm it's appropriate for a deployment to have to grant itself these perms in order to function? A fairly quick look and they seem fine to me, but I'm not an expert. I'll post a couple comments pointing to the specific perms I'm talking about.

@OndrejKotek thanks for this; it's very nicely done.

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/security/AuthenticationTestCase.java [hunk 1]
```diff
@@ -103,13 +107,31 @@
...
+                        new RuntimePermission("org.jboss.security.getSecurityContext"),
+                        new RuntimePermission("org.jboss.security.SecurityContextFactory.createSecurityContext"),
+                        new RuntimePermission("org.jboss.security.SecurityContextFactory.createUtil"),
+                        new RuntimePermission("org.jboss.security.plugins.JBossSecurityContext.setSubjectInfo"),
+                        new RuntimePermission("org.jboss.security.setSecurityContext"),
+                        // AuthenticationTestCase#execute calls ExecutorService#shutdownNow
+                        new RuntimePermission("modifyThread"),
+                        // AuthenticationTestCase#execute calls sun.net.www.http.HttpClient#openServer under the hood
+                        new SocketPermission(SERVER_HOST_PORT, "connect,resolve"),
+                        // TestSuiteEnvironment reads system properties
```
Comment (bstansberry) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/security/AuthenticationTestCase.java]: I'm looking for an expert check on all these ^^^. The ones below are fine.

File: testsuite/integration/xts/src/test/java/org/jboss/as/test/xts/annotation/client/TransactionalTestCase.java [hunk 2]
```diff
@@ -44,17 +49,33 @@
...
+                .addAsManifestResource(createPermissionsXmlAsset(
+                        // TransactionalClient#newInstance
+                        new RuntimePermission("getClassLoader"),
+                        new RuntimePermission("accessDeclaredMembers"),
+                        new FilePermission(javaHome + File.separator + "lib" + File.separator + "jaxws.properties", "read"), // javax.xml.ws.spi.FactoryFinder
+                        new FilePermission(javaHome + File.separator + "lib" + File.separator + "wsdl.properties", "read"),
+                        new PropertyPermission("user.dir", "read"),
+                        new RuntimePermission("org.apache.cxf.permission"),
+                        new SocketPermission(SERVER_HOST_PORT, "connect,resolve"),
+                        new RuntimePermission("createClassLoader"),
```
Comment (bstansberry) [inline testsuite/integration/xts/src/test/java/org/jboss/as/test/xts/annotation/client/TransactionalTestCase.java]: I'm looking for an expert check on all these ^^^. The ones below are fine.

**Comment by bstansberry** [comment]
The basic thing to check is whether the permission being added is needed to allow the container to provide some service to the deployment that is an expected part of the container's functionality. If so, that's a potential issue; it indicates the container code is lacking needed permissions. OTOH, if the permission is needed just to let some custom logic inside the deployment work, that's fine.

**Comment by stuartwdouglas** [comment]
With regard to the management client it looks like this is the correct fix: https://github.com/wildfly/wildfly-arquillian/commit/88faebf07f3c43095b58a582aab6106de550cca7

File: testsuite/integration/xts/src/test/java/org/jboss/as/test/xts/annotation/client/TransactionalTestCase.java (hunk 2 above)
Comment (stuartwdouglas) [inline testsuite/integration/xts/src/test/java/org/jboss/as/test/xts/annotation/client/TransactionalTestCase.java]: This does not look right. I tried running this locally and ran into https://github.com/wildfly/wildfly/pull/8121

After that was fixed I now get:

```
javax.xml.ws.WebServiceException: java.security.AccessControlException: WFSM000001: Permission check failed (permission "("java.lang.RuntimePermission" "getClassLoader")" in code source "(vfs:/content/transactional-test.war/WEB-INF/classes <no signer certificates>)" of "null")
	at org.wildfly.security.manager.WildFlySecurityManager.checkPermission(WildFlySecurityManager.java:273)
	at org.wildfly.security.manager.WildFlySecurityManager.checkPermission(WildFlySecurityManager.java:175)
	at java.lang.ClassLoader.checkClassLoaderPermission(ClassLoader.java:1525)
	at java.lang.Thread.getContextClassLoader(Thread.java:1436)
	at javax.xml.ws.spi.FactoryFinder.find(FactoryFinder.java:70)
	at javax.xml.ws.spi.Provider.provider(Provider.java:110)
	at javax.xml.ws.Service.<init>(Service.java:57)
	at javax.xml.ws.Service.create(Service.java:687)
	at org.jboss.as.test.xts.annotation.client.TransactionalClient.newInstance(TransactionalClient.java:57)
	at org.jboss.as.test.xts.annotation.client.TransactionalTestCase.testActiveTransaction(TransactionalTestCase.java:76)
```

Which is a problem with the javax.xml.ws.Service class (it should be doing its factory finder stuff in a PA). This is a real bug in our code when running under a security manager, and adding these permissions simply covers it up.

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/security/AuthenticationTestCase.java (hunk 1 above)
Comment (stuartwdouglas) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/ejb/security/AuthenticationTestCase.java]: These look ok. I think we should also consider adding the below properties to the security manager subsystem as part of the test suit setup. As they are clearly test suite specific their presence there will not mask real bugs.

**Comment by gytis** [comment]
Seems that pull request test with security manager didn't execute xts integration tests.
Locally I needed @stuartwdouglas fix for standalone-xts.xml to make it start the WildFly.

I only see getClassLoader permission issue that @stuartwdouglas mentioned, if RuntimePermission("getClassLoader") from this PR is removed.

However, this PR still fails for me locally and so far I've added two extra permissions:
FilePermission($WORK_DIR + "/xcatalog", "read") and RuntimePermission("getBootModuleLoader")). But now I get the following error:

```
com.arjuna.wst.SystemException: java.security.AccessControlException: WFSM000001: Permission check failed (permission "("java.io.FilePermission" "/Users/gytis/Workspace/wildfly/dist/target/wildfly-10.0.0.CR1-SNAPSHOT/modules/system/layers/base/org/jboss/xts/main/jbossxts-5.2.2.Final.jar" "read")" in code source "(vfs:/content/transactional-test.war/WEB-INF/classes <no signer certificates>)" of "null")
	at com.arjuna.mwlabs.wst11.at.remote.UserTransactionImple.startTransaction(UserTransactionImple.java:308)
	at com.arjuna.mwlabs.wst11.at.remote.UserTransactionImple.begin(UserTransactionImple.java:80)
	at com.arjuna.mwlabs.wst11.at.remote.UserTransactionImple.begin(UserTransactionImple.java:70)
	at org.jboss.as.test.xts.annotation.client.TransactionalTestCase.testActiveTransaction(TransactionalTestCase.java:104)
```

I'm going to look into this further, but if you have any suggestions give me a shout.

**Comment by gytis** [comment]
I've managed to make the test pass after adding following additional permissions:

```
new RuntimePermission("getBootModuleLoader"),
new FilePermission("<<ALL FILES>>", "read"),
new ReflectPermission("suppressAccessChecks"),
new RuntimePermission("com.arjuna.webservices11.ServiceRegistry.getRegistry"),
new RuntimePermission("setContextClassLoader")
```

**Comment by gytis** [comment]
@bstansberry , I think the permissions added by @OndrejKotek are ok to have in the test. However, those 5 I have mentioned before are needed to begin the transaction. Thus they probably should go to the subsystem. Do you have any suggestions how to people usually do that?

**Comment by bstansberry** [comment]
When the server is doing something for its own needs where it needs permissions it does that work in an AccessController.doPrivileged() block. The scope of that block needs to be carefully limited though. For example, within that block you need to watch out for calling out to end user code, thus letting that code execute with privileges.

**Comment by gytis** [comment]
By the way, only one of the XTS tests was updated here, but others should have missing permissions too. Was it intentional to select only TransactionalTestCase class?

**Comment by OndrejKotek** [comment]
There are other related issues, like https://issues.jboss.org/browse/WFLY-5170

**Comment by gytis** [comment]
I've raised the pull request for Narayana to add PrivilegedAction to many classes. This removes the need for all additional permissions I had to add to make this test work, except one. I still need FilePermission("<<ALL FILES>>", "read"), because otherwise the test fails to access org.jboss.xts and com.sun.xml.bind modules and wildfly/testsuite/integration/xts/target/workdir/xcatalog directory.

**Comment by OndrejKotek** [comment]
Fix for XTS tests removed. XTS fix is part of https://github.com/gytis/jboss-as/commit/6799626bff25cd40fe1f0a68e94b72c1bf1d67e7

**Comment by n1hility** [comment]
retest this please

**Comment by gytis** [comment]
@tomjenkinson , WildFly XTS tests fix was in our 5_BRANCH. Was that merged in?

**Pull request merged**